File: Freq_Filtry/MPI_Ach_Fch_Draw/Unit_AchFchParam.pas
unit Unit_AchFchParam;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, Buttons, ExtCtrls, Spin, Menus, ImgList;

type
  TfrmAchFchParam = class(TForm)
    Label1: TLabel;
    edBeginFreq: TEdit;
    Label2: TLabel;
    edEndFreq: TEdit;
    CheckBox2: TCheckBox;
    BitBtn1: TBitBtn;
    BitBtn2: TBitBtn;
    Label3: TLabel;
    Label4: TLabel;
    CheckBox4: TCheckBox;
    CheckBox5: TCheckBox;
    Label5: TLabel;
    Panel1: TPanel;
    Bevel2: TBevel;
    Label9: TLabel;
    Label10: TLabel;
    Panel2: TPanel;
    PaintBox2: TPaintBox;
    Label11: TLabel;
    SpinEdit2: TSpinEdit;
    Panel4: TPanel;
    Bevel1: TBevel;
    Label6: TLabel;
    Label7: TLabel;
    Panel3: TPanel;
    PaintBox1: TPaintBox;
    Label8: TLabel;
    SpinEdit1: TSpinEdit;
    Panel5: TPanel;
    ColorDialog1: TColorDialog;
    PopupMenu1: TPopupMenu;
    fg1: TMenuItem;
    N1: TMenuItem;
    N4: TMenuItem;
    N2: TMenuItem;
    N3: TMenuItem;
    PopupMenu2: TPopupMenu;
    MenuItem1: TMenuItem;
    MenuItem2: TMenuItem;
    MenuItem3: TMenuItem;
    MenuItem4: TMenuItem;
    MenuItem5: TMenuItem;
    ImageList1: TImageList;
    procedure BitBtn1Click(Sender: TObject);
    procedure Panel1Click(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure PaintBox1Paint(Sender: TObject);
    procedure PaintBox2Paint(Sender: TObject);
    procedure Panel5Click(Sender: TObject);
    procedure Panel4Click(Sender: TObject);
    procedure SpinEdit1Change(Sender: TObject);
    procedure SpinEdit2Change(Sender: TObject);
    procedure PaintBox1Click(Sender: TObject);
    procedure PaintBox2Click(Sender: TObject);
    procedure fg2(Sender: TObject);
    procedure MenuItem1Click(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  frmAchFchParam: TfrmAchFchParam;

implementation

uses Unit_Main, MPI_Library, MPI_Ach_Fch_Draw;

{$R *.DFM}

procedure TfrmAchFchParam.FormCreate(Sender: TObject);
begin
  AchFchParam.BeginFreq := StrToFloat(edBeginFreq.text);
  AchFchParam.EndFreq   := StrToFloat(edendFreq  .text);
  
  Panel1.Color                := Rgb(212, 208, 200);
  AchFchParam.BackGroundColor := Rgb(212, 208, 200);

  Panel5.Color         := clRed;
  AchFchParam.AchColor := clRed;

  Panel4.Color         := clYellow;
  AchFchParam.FchColor := clYellow;

  AchFchParam.AchPenWidth := 2;
  SpinEdit1.Value         := AchFchParam.AchPenWidth;
  AchFchParam.FchPenWidth := 2;
  SpinEdit2.Value         := AchFchParam.FchPenWidth;

  AchFchParam.AchPenStyle := psSolid;
  AchFchParam.FchPenStyle := psSolid;

  AchFchParam.AchEnabled := CheckBox4.Checked;
  AchFchParam.FchEnabled := CheckBox5.Checked;
  AchFchParam.t0Enabled  := CheckBox2.Checked;
end;

procedure TfrmAchFchParam.PaintBox1Paint(Sender: TObject);
var
 cy  : Integer;
begin
  cy := PaintBox1.Height div 2;
  With PaintBox1.Canvas do begin
     Brush.Color := Color;
     Pen.Color   := clBlack;
     Pen.Width   := 1;
     Pen.Style   := psSolid;
     Rectangle(0, 0, PaintBox1.Width, PaintBox1.Height);

     Pen.Color := AchFchParam.AchColor;
     Pen.Style := AchFchParam.AchPenStyle;
     Pen.Width := AchFchParam.AchPenWidth;
     MoveTo(2, cy - 1); LineTo(PaintBox1.Width-2, cy - 1);
     MoveTo(2, cy - 0); LineTo(PaintBox1.Width-2, cy - 0);
     MoveTo(2, cy + 1); LineTo(PaintBox1.Width-2, cy + 1);
  end;
end;

procedure TfrmAchFchParam.PaintBox2Paint(Sender: TObject);
var
 cy  : Integer;
begin
  cy := PaintBox2.Height div 2;
  With PaintBox2.Canvas do begin
     Brush.Color := Color;
     Pen.Color   := clBlack;
     Pen.Width   := 1;
     Pen.Style   := psSolid;
     Rectangle(0, 0, PaintBox2.Width, PaintBox2.Height);

     Pen.Color := AchFchParam.FchColor;
     Pen.Style := AchFchParam.FchPenStyle;
     Pen.Width := AchFchParam.FchPenWidth;
     MoveTo(2, cy - 1); LineTo(PaintBox2.Width-2, cy - 1);
     MoveTo(2, cy - 0); LineTo(PaintBox2.Width-2, cy - 0);
     MoveTo(2, cy + 1); LineTo(PaintBox2.Width-2, cy + 1);
  end;
end;

procedure TfrmAchFchParam.Panel1Click(Sender: TObject);
begin
  if ColorDialog1.Execute then begin
     Panel1.Color                := ColorDialog1.Color;
     AchFchParam.BackGroundColor := ColorDialog1.Color;
  end;
end;

procedure TfrmAchFchParam.Panel5Click(Sender: TObject);
begin
  if ColorDialog1.Execute then begin
     Panel5.Color         := ColorDialog1.Color;
     AchFchParam.AchColor := ColorDialog1.Color;
  end;
  PaintBox1.Repaint;
end;

procedure TfrmAchFchParam.Panel4Click(Sender: TObject);
begin
  if ColorDialog1.Execute then begin
     Panel4.Color         := ColorDialog1.Color;
     AchFchParam.FchColor := ColorDialog1.Color;
  end;
  PaintBox2.Repaint;
end;

procedure TfrmAchFchParam.SpinEdit1Change(Sender: TObject);
begin
  AchFchParam.AchPenWidth := SpinEdit1.Value;
end;

procedure TfrmAchFchParam.SpinEdit2Change(Sender: TObject);
begin
  AchFchParam.FchPenWidth := SpinEdit2.Value;
end;

procedure TfrmAchFchParam.PaintBox1Click(Sender: TObject);
var
  CurPoint    : TPoint;
begin
      CurPoint.X := Panel3.Left;
      CurPoint.Y := Panel3.Top + Panel3.Height+1;
      CurPoint   := ClientToScreen(CurPoint);
      PopUpMenu1.Popup(CurPoint.X, CurPoint.Y);
end;

procedure TfrmAchFchParam.PaintBox2Click(Sender: TObject);
var
  CurPoint    : TPoint;
begin
      CurPoint.X := Panel2.Left;
      CurPoint.Y := Panel2.Top + Panel2.Height+1;
      CurPoint   := ClientToScreen(CurPoint);
      PopUpMenu2.Popup(CurPoint.X, CurPoint.Y);
end;

procedure TfrmAchFchParam.fg2(Sender: TObject);
begin
  Case TComponent(Sender).tag of
    0 : AchFchParam.AchPenStyle := psSolid;
    1 : AchFchParam.AchPenStyle := psDash;
    2 : AchFchParam.AchPenStyle := psDot;
    3 : AchFchParam.AchPenStyle := psDashDot;
    4 : AchFchParam.AchPenStyle := psDashDotDot;
  end;
  if (AchFchParam.AchPenWidth <> 1) and (AchFchParam.AchPenStyle <> psSolid) then AchFchParam.AchPenWidth := 1;

  SpinEdit1.Value := AchFchParam.AchPenWidth;
  PaintBox1.Repaint;
end;

procedure TfrmAchFchParam.MenuItem1Click(Sender: TObject);
begin
  Case TComponent(Sender).tag of
    0 : AchFchParam.FchPenStyle := psSolid;
    1 : AchFchParam.FchPenStyle := psDash;
    2 : AchFchParam.FchPenStyle := psDot;
    3 : AchFchParam.FchPenStyle := psDashDot;
    4 : AchFchParam.FchPenStyle := psDashDotDot;
  end;
  if (AchFchParam.FchPenWidth <> 1) and (AchFchParam.FchPenStyle <> psSolid) then AchFchParam.FchPenWidth := 1;

  SpinEdit2.Value := AchFchParam.FchPenWidth;
  PaintBox2.Repaint;
end;

procedure TfrmAchFchParam.BitBtn1Click(Sender: TObject);
var
  Finish : Boolean;
begin
  Finish := True;

  try AchFchParam.BeginFreq := StrToFloat(edBeginFreq.text) except ShowWrongEdit(edBeginFreq); Finish := False; end;
  try AchFchParam.EndFreq   := StrToFloat(edendFreq  .text) except ShowWrongEdit(edEndFreq  ); Finish := False; end;

  AchFchParam.AchEnabled := CheckBox4.Checked;
  AchFchParam.FchEnabled := CheckBox5.Checked;
  AchFchParam.t0Enabled  := CheckBox2.Checked;

  if Finish then begin
     Ach_Fch_Draw();
     close
  end;
end;

end.

File: ach_fch_draw/Ach_Fch_Form/Unit_AchFchParam.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Unit_AchFchParam.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "CSPIN"
#pragma link "cdiroutl"
#pragma link "FloatEdit"
#pragma link "FloatEdit"
#pragma resource "*.dfm"
TfrmAchFchParam *frmAchFchParam;
//---------------------------------------------------------------------------
__fastcall TfrmAchFchParam::TfrmAchFchParam(TComponent* Owner) : TForm(Owner)
{
  ParentForm = (TfrmAchFch*) Owner;
}
//---------------------------------------------------------------------------
void __fastcall TfrmAchFchParam::FormShow(TObject *Sender)
{
   Panel1->Color = ParentForm->AchFchBackGroundColor;
   Panel5->Color = ParentForm->AchColor;
   Panel4->Color = ParentForm->FchColor;
   CSpinEdit1->Value = ParentForm->AchPenWidth;
   CSpinEdit2->Value = ParentForm->FchPenWidth;

   FloatEdit1->Set_Float_Value(ParentForm->BeginFreq);
   FloatEdit2->Set_Float_Value(ParentForm->EndFreq);

    CheckBox4->Checked = ParentForm->AchEnabled;
    CheckBox5->Checked = ParentForm->FchEnabled;
    CheckBox2->Checked = ParentForm->Enable_Label_T0;
    CheckBox1->Checked = ParentForm->RamkaEnabled;
}
//---------------------------------------------------------------------------

void __fastcall TfrmAchFchParam::BitBtn1Click(TObject *Sender)
{
   bool Res = true; 

   FloatEdit1->Get_Float_Value(&ParentForm->BeginFreq, &Res);
   FloatEdit2->Get_Float_Value(&ParentForm->EndFreq  , &Res);

  ParentForm->AchEnabled       = CheckBox4->Checked;
  ParentForm->FchEnabled       = CheckBox5->Checked;
  ParentForm->Enable_Label_T0  = CheckBox2->Checked;
  ParentForm->RamkaEnabled     = CheckBox1->Checked;

  if (Res) Close();
}
//---------------------------------------------------------------------------


void __fastcall TfrmAchFchParam::PaintBox1Paint(TObject *Sender)
{
     int cy = PaintBox1->Height / 2;

     PaintBox1->Canvas->Brush->Color = Color;
     PaintBox1->Canvas->Pen->Color   = clBlack;
     PaintBox1->Canvas->Pen->Width   = 1;
     PaintBox1->Canvas->Pen->Style   = psSolid;
     PaintBox1->Canvas->Rectangle(0, 0, PaintBox1->Width, PaintBox1->Height);

     PaintBox1->Canvas->Pen->Color = ParentForm->AchColor;
     PaintBox1->Canvas->Pen->Style = ParentForm->AchPenStyle;
     PaintBox1->Canvas->Pen->Width = ParentForm->AchPenWidth;
     PaintBox1->Canvas->MoveTo(2, cy - 1); PaintBox1->Canvas->LineTo(PaintBox1->Width-2, cy - 1);
     PaintBox1->Canvas->MoveTo(2, cy - 0); PaintBox1->Canvas->LineTo(PaintBox1->Width-2, cy - 0);
     PaintBox1->Canvas->MoveTo(2, cy + 1); PaintBox1->Canvas->LineTo(PaintBox1->Width-2, cy + 1);
}
//---------------------------------------------------------------------------

void __fastcall TfrmAchFchParam::PaintBox2Paint(TObject *Sender)
{
     int cy = PaintBox1->Height / 2;

     PaintBox2->Canvas->Brush->Color = Color;
     PaintBox2->Canvas->Pen->Color   = clBlack;
     PaintBox2->Canvas->Pen->Width   = 1;
     PaintBox2->Canvas->Pen->Style   = psSolid;
     PaintBox2->Canvas->Rectangle(0, 0, PaintBox1->Width, PaintBox1->Height);

     PaintBox2->Canvas->Pen->Color = ParentForm->FchColor;
     PaintBox2->Canvas->Pen->Style = ParentForm->FchPenStyle;
     PaintBox2->Canvas->Pen->Width = ParentForm->FchPenWidth;
     PaintBox2->Canvas->MoveTo(2, cy - 1); PaintBox2->Canvas->LineTo(PaintBox2->Width-2, cy - 1);
     PaintBox2->Canvas->MoveTo(2, cy - 0); PaintBox2->Canvas->LineTo(PaintBox2->Width-2, cy - 0);
     PaintBox2->Canvas->MoveTo(2, cy + 1); PaintBox2->Canvas->LineTo(PaintBox2->Width-2, cy + 1);
}
//---------------------------------------------------------------------------

void __fastcall TfrmAchFchParam::Panel1Click(TObject *Sender)
{
  ColorDialog1->Color = ParentForm->AchFchBackGroundColor;
  if (ColorDialog1->Execute()){
     if (ColorDialog1->Color == clBlack) return;
     Panel1->Color                     = ColorDialog1->Color;
     ParentForm->AchFchBackGroundColor = ColorDialog1->Color;
  }

}
//---------------------------------------------------------------------------

void __fastcall TfrmAchFchParam::Panel5Click(TObject *Sender)
{
  ColorDialog1->Color = ParentForm->AchColor;
  if (ColorDialog1->Execute()){
     if (ColorDialog1->Color == clWhite) return;
     Panel5->Color        = ColorDialog1->Color;
     ParentForm->AchColor = ColorDialog1->Color;
  }
  PaintBox1->Repaint();
}
//---------------------------------------------------------------------------

void __fastcall TfrmAchFchParam::Panel4Click(TObject *Sender)
{
  ColorDialog1->Color = ParentForm->FchColor;
  if (ColorDialog1->Execute()){
     if (ColorDialog1->Color == clWhite) return;
     Panel4->Color        = ColorDialog1->Color;
     ParentForm->FchColor = ColorDialog1->Color;
  }
  PaintBox2->Repaint();
}
//---------------------------------------------------------------------------

void __fastcall TfrmAchFchParam::CSpinEdit1Change(TObject *Sender)
{
  ParentForm->AchPenWidth = CSpinEdit1->Value;

}
//---------------------------------------------------------------------------

void __fastcall TfrmAchFchParam::CSpinEdit2Change(TObject *Sender)
{
  ParentForm->FchPenWidth = CSpinEdit2->Value;

}
//---------------------------------------------------------------------------

void __fastcall TfrmAchFchParam::PaintBox1Click(TObject *Sender)
{
  TPoint  CurPoint;

      CurPoint.x = Panel3->Left;
      CurPoint.y = Panel3->Top + Panel3->Height+1;
      CurPoint   = ClientToScreen(CurPoint);
      PopupMenu1->Popup(CurPoint.x, CurPoint.y);
}
//---------------------------------------------------------------------------

void __fastcall TfrmAchFchParam::PaintBox2Click(TObject *Sender)
{
  TPoint  CurPoint;

      CurPoint.x = Panel2->Left;
      CurPoint.y = Panel2->Top + Panel2->Height+1;
      CurPoint   = ClientToScreen(CurPoint);
      PopupMenu2->Popup(CurPoint.x, CurPoint.y);
}
//---------------------------------------------------------------------------

void __fastcall TfrmAchFchParam::fg1Click(TObject *Sender)
{
  switch (((TComponent*)Sender)->Tag) {
    case 0 : ParentForm->AchPenStyle = psSolid;      break;
    case 1 : ParentForm->AchPenStyle = psDash;       break;
    case 2 : ParentForm->AchPenStyle = psDot;        break;
    case 3 : ParentForm->AchPenStyle = psDashDot;    break;
    case 4 : ParentForm->AchPenStyle = psDashDotDot; break;
  }
  if (ParentForm->AchPenWidth != 1 && ParentForm->AchPenStyle != psSolid) ParentForm->AchPenWidth = 1;

  CSpinEdit1->Value = ParentForm->AchPenWidth;
  PaintBox1->Repaint();
}
//---------------------------------------------------------------------------

void __fastcall TfrmAchFchParam::MenuItem1Click(TObject *Sender)
{
  switch (((TComponent*)Sender)->Tag) {
    case 0 : ParentForm->FchPenStyle = psSolid;      break;
    case 1 : ParentForm->FchPenStyle = psDash;       break;
    case 2 : ParentForm->FchPenStyle = psDot;        break;
    case 3 : ParentForm->FchPenStyle = psDashDot;    break;
    case 4 : ParentForm->FchPenStyle = psDashDotDot; break;
  }
  if (ParentForm->FchPenWidth != 1 && ParentForm->FchPenStyle != psSolid) ParentForm->FchPenWidth = 1;

  CSpinEdit2->Value = ParentForm->FchPenWidth;
  PaintBox2->Repaint();
}
//---------------------------------------------------------------------------





